_posts/2016-05-11-android-testing-with-kotlin.md

```diff
@@ -0,0 +1,422 @@
+project's Gradle build file, declaring which version of Kotlin you'd like to use and
+which dependencies need to be additionally compiled. As far my experience goes,
+[the official documentation](https://kotlinlang.org/docs/reference/using-gradle.html),
+which seem to be great, will point you in the right direction. In any case, my `build.gradle`
+build script ended up being like this:
+
+```groovy
+apply plugin: 'com.android.application'
+apply plugin: 'kotlin-android'
+
+buildscript {
+    ext.kotlin_version = '1.0.+'
+    repositories {
+        jcenter()
+        mavenCentral()
+    }
+    dependencies {
+        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
+    }
+}
+
+android {
+    ...
+}
+
+dependencies {
+    compile "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+    testCompile "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+    testCompile "org.jetbrains.kotlin:kotlin-test-junit:$kotlin_version"
+    ...
+}
+```
+
+Let us tear down the extra lines. On the second line, right after Android plugin is applied,
+we inject Kotlin Gradle plugin that targets Android build model by
+`apply plugin: 'kotlin-android'`.
+To make that plugin available we need to declare it as a `buildscript` level dependency with
+
+```groovy
+dependencies {
+    classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
+}
+```
+where the preceding `ext.kotlin_version = '1.0.+'` just selects the newest version of Kotlin from
+1.0 series.
+
+Finally, under `dependencies` section we need to state that Kotlin _stdlib_ has to be compiled.
+And as we are going to use Kotlin especially for testing purposes, it needs to be compiled for
+tests as well.
+
+__One note about tests and their location in the source tree.__
+
+As unit and UI tests are usually stored in `src/test/java` and `src/androidTest/java`
+respectively, we might want to maintain sanity in our codebase and keep Kotlin tests
+in dedicated directories like `src/test/kotlin` and `src/androidTest/kotlin`. As these
+are directories are not detected as Android test locations by default (*yet?*), we
+need to make Gradle and Android Studio aware of them. This can be done in the `sourceSets`
+section with
+
+```groovy
+android {
+    ...
+    sourceSets {
+        test.java.srcDirs += 'src/test/kotlin'
+        androidTest.java.srcDirs += 'src/androidTest/kotlin'
+    }
+}
+```
+
+## Our example app
+
+The [application](https://github.com/GreenhouseCI/kotlin-tests) created for
+the purposes of this blog post is of course yet another stunning shopping list app. As expected,
+it is possible to add some items to the list and remove them from there as you wish. To add
+some spice into the mix and make things a little bit more __testable__, every item that is
+inserted to the list passes through a formatter that removes all excess whitespace from the
+input string. Critical business logic in the form of our formatter that we use to clear up
```

Comment:
My extensive use or *our/we* probably originates from the math papers I've read. For some reason mathematicians always tend to include reader using phrases in the form of *this equation we have*, *our proof for*, etc.